### ISSUE TITLE
How to use early stop when training VGG16?

### ISSUE THREAD

**Issue by PipiZong** [tianyic/only_train_once#3]
Hi, thanks for sharing your great work. Just wondering if we have a validation set, how can we use it for an early stop when training the VGG16 model in your case? I noticed you trained all epochs and saved the checkpoint of the last epoch as the final model for pruning.

**Comment by tianyic** [comment]
Thanks for the question. 

We would suggest to early stop if the group sparsity becomes stationary and the validation accuracy becomes satisfactory.

Once group sparsity is stationary, the checkpoint with the best validation accuracy serves as the better checkpoint to return than the last checkpoint. We will revise the logic in the next release of OTO framework.

**Comment by PipiZong** [comment]
Hi,

Thanks for your reply. Sorry, another question, how to apply OTO to the embedding layers? Will it be the same as applying in MLPs?
I am looking forward to your reply.

**Comment by tianyic** [comment]
It depends, if the embedding layers are prescribed, i.e., they are not trainable, ZIG partition would skip these parameters. Otherwise, if the embedding layers are trainable, the ZIG should perform similarly to the fully connected or linear layers.

**Comment by tianyic** [comment]
In particular, ZIG is essentially the minimal removal structure of DNN. If applying onto trainable embedding layer, we need to remove the same amount scalars from all embedding vectors to keep the followup operators valid.

If we stack all embedding vectors as an embedding matrix, where each column is one embedding vector, then similarly to the linear layer, each row serves as one ZIG. 

Let me know if any further question.
